shell/browser/api/electron_api_base_window.cc

```diff
@@ -177,8 +177,15 @@ void BaseWindow::OnWindowClosed() {
       FROM_HERE, GetDestroyClosure());
 }
 
-void BaseWindow::OnWindowEndSession() {
-  Emit("session-end");
+void BaseWindow::OnWindowQueryEndSession(const std::vector<std::string> reasons,
+                                         bool* prevent_default) {
+  if (Emit("query-session-end", reasons)) {
+    *prevent_default = true;
+  }
```

Comment:
We have a newer pattern where we expose the values on the Event object directly. Can you change this event and the other to use this approach? Here's what it'd look like here: 
```suggestion
  v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
  v8::HandleScope handle_scope(isolate);
  gin::Handle<gin_helper::internal::Event> event =
      gin_helper::internal::Event::New(isolate);
  v8::Local<v8::Object> event_object = event.ToV8().As<v8::Object>();
  gin::Dictionary dict(isolate, event_object);
  dict.Set("reasons", reasons);
  EmitWithoutEvent("query-session-end", event);
  if (event->GetDefaultPrevented()) {
    *prevent_default = true;
  }
```

To see how to document this, take a look at the changes to expose the 'audio-state-changed' event: https://github.com/electron/electron/commit/512e56baf75bce86aef11184af881cff67d8e6a2

Reply:
@samuelmaddock I tried to document it the same way, but unfortunately I am getting a lot of errors during `electron-api.json` generation:
```
../electron.d.ts(3898,5): error TS2416: Property 'on' in type 'BrowserWindow' is not assignable to the same property in base type 'BaseWindow'.
  Type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "always-on-top-changed", listener: (event: { ...; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, com...' is not assignable to type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, command: string) => void): this; (event: "blur", listener: (event: { ...; }) => void): this; (event: "cl...'.
    Types of parameters 'event' and 'event' are incompatible.
      Type '"query-session-end"' is not assignable to type '"always-on-top-changed"'.
../electron.d.ts(3900,5): error TS2416: Property 'off' in type 'BrowserWindow' is not assignable to the same property in base type 'BaseWindow'.
  Type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "always-on-top-changed", listener: (event: { ...; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, com...' is not assignable to type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, command: string) => void): this; (event: "blur", listener: (event: { ...; }) => void): this; (event: "cl...'.
    Types of parameters 'event' and 'event' are incompatible.
      Type '"query-session-end"' is not assignable to type '"always-on-top-changed"'.
../electron.d.ts(3902,5): error TS2416: Property 'once' in type 'BrowserWindow' is not assignable to the same property in base type 'BaseWindow'.
  Type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "always-on-top-changed", listener: (event: { ...; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, com...' is not assignable to type '{ (event: "always-on-top-changed", listener: (event: { preventDefault: () => void; readonly defaultPrevented: boolean; }, isAlwaysOnTop: boolean) => void): this; (event: "app-command", listener: (event: { ...; }, command: string) => void): this; (event: "blur", listener: (event: { ...; }) => void): this; (event: "cl...'.
    Types of parameters 'event' and 'event' are incompatible.
      Type '"query-session-end"' is not assignable to type '"always-on-top-changed"'.
[...]
```

I would suggest it's because all BaseWindow events should be ported in the same way.

Reply:
I forget escaping, sorry 🤦

Reply:
I found the root of problem: after adding event to both `base-window.md` and `browser-window.md` it will generate four types:
- BaseWindowQuerySessionEndEventParams
- BaseWindowSessionEndEventParams
- BrowserWindowQuerySessionEndEventParams
- BrowserWindowSessionEndEventParams

After manually removing two latter types from `electron.d.td` it started to work, but I don't know how to fix it without fixing typing generator.

Reply:
This does unfortunately seem like something that needs to be fixed with our type generator. We want BrowserWindow docs on the website to show everything in BaseWindow so we've been duplicating the docs for now. A real fix might involve changing how our website is generated to include inherited methods.

For the purpose of this PR though, I think we could just have these types in BrowserWindow so that it's included in the BrowserWindow docs on the website as well. I'd be curious if other folks in @electron/wg-api have thoughts on how to address this.

Reply:
What if you define a structure for those events, and reference the same structure(s) from both `BrowserWindow` and `BaseWindow`?

An example of such a structure that extends Event can be found here: https://github.com/electron/electron/blob/main/docs/api/structures/ipc-main-event.md

If you did that, then maybe TypeScript would consider the types compatible?

Reply:
@samuelmaddock there is another problem. I am not C++ expert, but `BaseWindow` C++ class inherits `gin_helper::TrackableObject` which seems like cannot be used with `gin_helper::EventEmitterMixin` which I need to use `EmitWithoutEvent`.

My guess is that probably `BaseWindow` class should be fully refactored to be compatible with `EventEmitterMixin`.

Reply:
I pushed a very naive fix, it compiles properly, but I am not sure this is the way.